### DISCUSSION TITLE
Executing a system once during runtime from another system

### DISCUSSION THREAD

**Discussion by andreashgk** [bevyengine/bevy#3112] [Q&A] [answered]
I am currently working on a simple prototype to experiment with Bevy
So far I am really enjoying how simple everything is, even tho this is my first rust project ever Bevy makes it pretty intuitive

However I am unsure how to do the following:
I want to make it so when I presses `O` on their keyboard it will spawn a dummy player (for testing).
I have this system to listen to keypressed:
```rust
pub fn debug_spawn(input: Res<Input<KeyCode>>) {
    if input.just_pressed(KeyCode::O) {
        // somehow execute spawn_dummy
    }
}
```
And the following system to spawn a dummy player:
```rust
pub fn spawn_dummy(mut commands: Commands, materials: Res<Materials>, mut c_mat: ResMut<Assets<ColorMaterial>>) {
    let mut mat = materials.player_material.clone();
    c_mat.get_mut(&mat).unwrap().color = AlignmentColors::Enemy.into();

    let sprite = SpriteBundle {
        material: mat,
        ..Default::default()
    };

    commands.spawn_bundle(sprite).insert_bundle(PlayerBundle {
        ..Default::default()
    });
}
```

How can I make the `spawn_dummy` system run from the `debug_spawn` system? Do I just add `mut commands: Commands, materials: Res<Materials>, mut c_mat: ResMut<Assets<ColorMaterial>>` as parameters to debug_spawn and call spawn_dummy manually, or is there a better way to do this?

**Comment by hymm** [comment, marked as answer]
you're probably looking for events https://bevy-cheatbook.github.io/programming/events.html

**Reply by andreashgk** [reply]
Thanks, I'll take a look!

**Reply by djeedai** [reply]
Note that if you're sure you want that to happen only once it's probably better to call `spawn_dummy()` directly, to avoid leaving registered a system (see #3131). But here if you do it each time a key is pressed, this is not "once" but "once in a while", and events are indeed better suited for that kind of thing, to decouple a bit the two systems.

**Comment by tatref** [comment]
As of 0.12, you probably want [One-Shot Systems](https://bevyengine.org/news/bevy-0-12/#one-shot-systems)
